### PR TITLE
Handle foreign key violations on delete resource [DB-level validation]

### PR DESCRIPTION
**Pull request by NickyMateev**
# Handle foreign key violations on delete resource [DB-level validation]

## Motivation [related to #360]

Delete requests which fail with foreign key violation should return Bad Request 400 instead of 
Internal Server Error 500.

## Solution

**DELETE /v1/platforms/<platform_id>**

Response Status: _400 Bad Request_
```json
{
  "error": "ExistingReferenceEntity",
  "description": "Error deleting entity Platform: Could not delete entity Platform due to existing reference entity ServiceInstance"
}
```

## Pros/Cons of such a DB-level validation approach

Pros:
* Central for all entities
* No extra DB calls
* No race conditions (like in API-level validation)

Cons:
* Cannot have custom pluggable API-level validation
* Cannot fetch IDs of violation entities (once transaction receives a foreign key constraint violation error, the transaction is marked as unusable and we cannot make any further queries to fetch the violation IDs)

## Pull Request status

* [x] Initial implementation
* [x] Integration tests

### CODE REVIEW

File: storage/postgres/storage.go [hunk 1]
```diff
@@ -245,6 +246,13 @@ func (ps *Storage) Delete(ctx context.Context, objType types.ObjectType, criteri
 	rows, err := ps.queryBuilder.NewQuery(entity).WithCriteria(criteria...).Return("*").Delete(ctx)
 	defer closeRows(ctx, rows)
 	if err != nil {
+		pqError, ok := err.(*pq.Error)
+		if ok && pqError.Code.Name() == "foreign_key_violation" {
```
Comment (pankrator) [inline storage/postgres/storage.go]: Extract constant "foreign_key_violation" or I think there are predefined code names in pq package

Comment (NickyMateev) [reply storage/postgres/storage.go]: We make use of the _ foreign_key_violation_ string, as this is what pq returns: https://github.com/lib/pq/blob/f91d3411e481ed313eeab65ebfe9076466c39d01/error.go#L177

Unfortunately, beyond a plain a string value to compare with, they don't have an explicit error type which we can assert, to ensure for such an error.

**Comment by dpanayotov** [comment]
I'd really like to have a user-friendly message. Like "I can't delete the platform, because there's instance with id <id> associated with it". Not sure if we can achieve this actually, but the current error is not much better that the one from the db.

**Comment by NickyMateev** [comment]
> I'd really like to have a user-friendly message. Like "I can't delete the platform, because there's instance with id associated with it". Not sure if we can achieve this actually, but the current error is not much better that the one from the db.

Discussed it with @KirilKabakchiev  and @pankrator , will provide a different approach soon.

**Comment by NickyMateev** [comment]
New representation:

```json
{
  "error": "ExistingReferenceEntity",
  "description": "Error deleting entity Platform: Could not delete entity Platform due to existing reference entity ServiceInstance"
}
```

File: test/broker_test/broker_test.go [hunk 2]
```diff
@@ -1115,33 +1115,86 @@ var _ = test.DescribeTestsFor(test.TestCase{
...
+
+									serviceInstanceIDs = append(serviceInstanceIDs, serviceInstance.ID)
+								}
+
+								ctx.SMWithOAuth.PATCH(web.ServiceBrokersURL + "/" + brokerID).
+									WithJSON(common.Object{}).
+									Expect().
+									Status(http.StatusBadRequest).
+									JSON().Object().
+									Value("description").String().Contains("existing reference entity")
```
Comment (dotchev) [inline test/broker_test/broker_test.go]: better check for error type "ExistingReferenceEntity"

File: test/broker_test/broker_test.go [hunk 3]
```diff
@@ -1115,33 +1115,86 @@ var _ = test.DescribeTestsFor(test.TestCase{
...
+
+								serviceOfferings := ctx.SMWithOAuth.ListWithQuery(web.ServiceOfferingsURL, fmt.Sprintf("fieldQuery=id eq '%s'", serviceOfferingID))
```
Comment (dotchev) [inline test/broker_test/broker_test.go]: instead of listing, just fetch it with the id

File: test/broker_test/broker_test.go [hunk 4]
```diff
@@ -1115,33 +1115,86 @@ var _ = test.DescribeTestsFor(test.TestCase{
...
+								serviceOfferings.NotEmpty()
+
+								servicePlans := ctx.SMWithOAuth.ListWithQuery(web.ServicePlansURL, "fieldQuery="+fmt.Sprintf("id in ('%s')", strings.Join(planIDsForService, "','")))
+								servicePlans.NotEmpty()
+								servicePlans.Length().Equal(len(planIDsForService))
+
+								assertInvocationCount(brokerServer.CatalogEndpointRequests, 1)
```
Comment (dotchev) [inline test/broker_test/broker_test.go]: why is that necessary?

File: test/test.go [hunk 5]
```diff
@@ -157,3 +159,45 @@ func DescribeTestsFor(t TestCase) bool {
 		}()
 	})
 }
+
+func PrepareServiceInstance(ctx *common.TestContext, auth *common.SMExpect, platformID, planID string, OSBContext string) (string, *types.ServiceInstance) {
```
Comment (dotchev) [inline test/test.go]: this function does not fit to the scope of this file

File: api/base_controller.go [hunk 6]
```diff
@@ -394,3 +395,7 @@ func pageFromObjectList(ctx context.Context, objectList types.ObjectList, count,
 	}
 	return page
 }
+
```
[the remark this answers is not in the document]
Comment (NickyMateev) [reply api/base_controller.go]: Yep, I actually implemented the `String()` interface for the ObjectType and I think it's much clearer now.

File: api/base_controller.go [hunk 7]
```diff
@@ -137,7 +138,7 @@ func (c *BaseController) CreateObject(r *web.Request) (*web.Response, error) {
 
 	createdObj, err := c.repository.Create(ctx, result)
 	if err != nil {
-		return nil, util.HandleStorageError(err, string(c.objectType))
+		return nil, util.HandleStorageError(err, trimObjectTypePrefix(c.objectType))
```
[the remark this answers is not in the document]
Comment (NickyMateev) [reply api/base_controller.go]: Addressed this by implementing the String interface

File: pkg/util/errors.go [hunk 8]
```diff
@@ -113,6 +112,16 @@ func (e *ErrBadRequestStorage) Error() string {
...
 }
 
+// ErrExistingReferenceEntityStorage represents a foreign key constraint storage error that should be translated to a user-friendly http.StatusBadRequest
+type ErrExistingReferenceEntityStorage struct {
+	Entity          string
+	ViolationEntity string
+}
+
+func (e *ErrExistingReferenceEntityStorage) Error() string {
+	return fmt.Sprintf("Could not delete entity %s due to existing reference entity %s", e.Entity, e.ViolationEntity)
```
[the remark this answers is not in the document]
Comment (NickyMateev) [reply pkg/util/errors.go]: Depends on the `pq` library and whatever it spits out, it's not really in our control, but it should always tell us of some entity which is the main root of us not being able to delete the original entity

File: pkg/util/errors.go [hunk 9]
```diff
@@ -149,6 +158,12 @@ func HandleStorageError(err error, entityName string) error {
 	default:
 		// in case we did not replace the pg.Error in the DB layer, propagate it as response message to give the caller relevant info
 		switch e := err.(type) {
+		case *ErrExistingReferenceEntityStorage:
+			return &HTTPError{
+				ErrorType:   "ExistingReferenceEntity",
+				Description: fmt.Sprintf("Error deleting entity %s: %s", entityName, err.Error()),
```
[the remark this answers is not in the document]
Comment (NickyMateev) [reply pkg/util/errors.go]: This message behaviour is intentional.

Yes - it seems redundant when you look at the case when you try to delete a platform, but can't because of an existing instance: 

> "Error deleting entity **Platform**: Could not delete entity **Platform** due to existing reference entity **ServiceInstance**"

But if you try to delete a Broker, this message will look like this:

> "Error deleting entity **Broker**: Could not delete entity **ServicePlan** due to existing reference entity **ServiceInstance**"

So in cases where there are entities between the original entity and violation (reference) entity, like it is here with the **ServicePlan** entity, it's a much more useful message, as you get a bigger picture of what's going on. If I were to remove the part that you request, the message will even be strange in the case I want to delete a **Broker** with existing instances:

> "Could not delete entity **ServicePlan** due to existing reference entity **ServiceInstance**"

File: api/base_controller.go [hunk 10]
```diff
@@ -394,3 +395,7 @@ func pageFromObjectList(ctx context.Context, objectList types.ObjectList, count,
 	}
 	return page
 }
+
+func trimObjectTypePrefix(objectType types.ObjectType) string {
+	return strings.TrimPrefix(string(objectType), types.Prefix)
```
[the remark this answers is not in the document]
Comment (NickyMateev) [reply api/base_controller.go]: Not sure I understand your question

[the remark this answers is not in the document]
Comment (NickyMateev) [reply api/base_controller.go]: When converting from EntityType(storage entity) to ObjectType(api entity) with the `ToObjectType()` method I add the prefix, yes, because all of our ObjectType values have this prefix.

I think it would be lead to a lot of inconsistencies and possibly bugs if we start relying on this `ToObjectType()` function later on and expect it to output "standard" `ObjectType` entities, but it does not.

File: test/test.go (hunk 5 above)
Comment (NickyMateev) [reply test/test.go]: It doesn't really, I agree. Do you have any better suggestions?

File: test/broker_test/broker_test.go (hunk 4 above)
Comment (NickyMateev) [reply test/broker_test/broker_test.go]: it's not necessary, I've removed it.

File: storage/postgres/storage.go [hunk 11]
```diff
@@ -245,6 +250,17 @@ func (ps *Storage) Delete(ctx context.Context, objType types.ObjectType, criteri
 	rows, err := ps.queryBuilder.NewQuery(entity).WithCriteria(criteria...).Return("*").Delete(ctx)
 	defer closeRows(ctx, rows)
 	if err != nil {
+		pqError, ok := err.(*pq.Error)
+		if ok && pqError.Code.Name() == foreignKeyViolation {
+			entityName := string(entity.GetType().ToObjectType())
```
[the remark this answers is not in the document]
Comment (NickyMateev) [reply storage/postgres/storage.go]: Good catch!

File: storage/interfaces.go [hunk 12]
```diff
@@ -126,6 +130,20 @@ func (s *NotificationSettings) Validate() error {
...
+	name = name[:len(name)-1]
+	if !strings.ContainsRune(name, '_') {
+		return types.ObjectType(types.Prefix + strings.Title(name))
+	}
+
+	name = strings.Replace(name, "_", " ", -1)
+	name = strings.Title(name)
+	name = strings.Replace(name, " ", "", -1)
+	return types.ObjectType(types.Prefix + name)
+}
```
[the remark this answers is not in the document]
Comment (NickyMateev) [reply storage/interfaces.go]: Wouldn't this mean that in order to introduce a new entity to SM we would not only have to implement the StorageEntity and then introduce it with `Intrduce()`, but we would also not have to forget to add a new pair of <storage type> - <object/api type> to this map?

File: cmd/smgen/storage_type_template.go [hunk 13]
```diff
@@ -36,14 +36,18 @@ import (
 
 var _ {{.StoragePackage}}PostgresEntity = &{{.Type}}{}
 
-const {{.Type}}Table = "{{.TableName}}"
+const {{.Type}}Table storage.EntityType = "{{.TableName}}"
+
+func (*{{.Type}}) GetType() storage.EntityType {
+	return {{.Type}}Table
```
[the remark this answers is not in the document]
Comment (NickyMateev) [reply cmd/smgen/storage_type_template.go]: No, it has to reference the corresponding `EntityType` constant defined for each storage Entity at the top of each `..._gen.go` file.

File: cmd/smgen/storage_type_template.go [hunk 14]
```diff
@@ -36,14 +36,18 @@ import (
...
+}
 
 func (*{{.Type}}) LabelEntity() {{.StoragePackage}}PostgresLabel {
 	return &{{.Type}}Label{}
 }
 
-func (*{{.Type}}) TableName() string {
-	return {{.Type}}Table
+func (e *{{.Type}}) TableName() string {
```
[the remark this answers is not in the document]
Comment (NickyMateev) [reply cmd/smgen/storage_type_template.go]: Seems like we don't need GetType() anymore, it was part of my initial implementation
